Add option to only cycle through populated desktops

givani30 opened the pull request:
## Summary
  This PR adds a new configuration option `cycle_populated_only` that allows users to cycle only through virtual desktops that contain windows.

  ## Changes
  - Add `isDeskPopulated()` method to check if a vdesk has any windows
  - Modify `nextDeskId()` and `prevDeskId()` to respect the new config option
  - Add `cycle_populated_only` configuration option with default value 0
  - Maintain backwards compatibility by defaulting to disabled

  ## Configuration
  ```ini
  plugin {
      virtual-desktops {
          cycle_populated_only = 0  # Default: cycle through all desktops
          # cycle_populated_only = 1  # Only cycle only through populated desktops
      }
  }
``` 

  Testing

  -  Default behavior (cycle_populated_only = 0): cycles through all virtual desktops
  -  New behavior (cycle_populated_only = 1): skips empty desktops
  -  Edge cases: handles single populated desktop and no populated desktops correctly

  Backward Compatibility

  This change is fully backward compatible - existing configurations will continue to work unchanged.

givani30 committed:
feat: add cycle_populated_only configuration option

Add new configuration option `cycle_populated_only` that allows users to cycle only through virtual desktops that contain windows.

- Add isDeskPopulated() method to check if a vdesk has any windows
- Modify nextDeskId() and prevDeskId() to respect the new config option
- Default value is 0 (disabled) to maintain backwards compatibility
- When enabled (1), cycling skips empty virtual desktops

givani30 committed:
refactor(ci): Simplify CI by installing Hyprland from packages

Instead of building Hyprland and its dependencies from source, this change simplifies the CI workflow to install them directly from the Arch Linux repositories using pacman.

This was motivated by recurring build failures when compiling Hyprland from source, making the CI process unreliable. Installing the pre-built packages is faster, more robust, and easier to maintain.

levnikmyskin commented on src/VirtualDeskManager.cpp:
```diff
@@ -252,24 +252,66 @@ void VirtualDeskManager::resetVdesk(const std::string& arg) {
+    bool populatedOnly = **PCYCLE_POPULATED_ONLY;
+
+    int currentId = activeVdesk()->id;
+    int candidateId = currentId;
```
shouldn't we just use `candidateId = activeVdesk()->id -1;`?

levnikmyskin commented on src/VirtualDeskManager.cpp:
```diff
@@ -252,24 +252,66 @@ void VirtualDeskManager::resetVdesk(const std::string& arg) {
+    int maxId = std::ranges::max(keys);
+    do {
+        candidateId--;
+        if (candidateId < minId) {
+            if (backwardCycle)
+                candidateId = maxId;
+            else
+                return minId;
         }
-    }
-    return prevId;
+        // A desk must exist to be a candidate
+        if (vdesksMap.contains(candidateId) && (!populatedOnly || isDeskPopulated(candidateId)))
+            return candidateId;
+    } while (candidateId != currentId);
```
I'm not a big fan of these lines tbh. We do `while(candidateId != currentId)`, but is that condition ever false?
Considering that `candidateId` should already be the activeVdesk - 1, then this can just be a `while(true)`. Or maybe better, a `while(candidateId >= minId)`, moving the `if(backwardCycle)` check after the while. Basically:
```cpp
while (candidateId >= minId) {
    if (!populatedOnly || isDeskPopulated(candidateId))
        return candidateId;
    candidateId--;
}

// If there's no other previous populated desk,
// we either return the maxId, if populated,
// or this is a no-op otherwise
if (populatedOnly) {
    if (backwardCycle && isDeskPopulated(maxId))
        return maxId;
    return activeVdesk()->id;
}

return backwardCycle ? maxId : minId;
```
I also removed the check on the vdesksmap, as that was not done before and I don't believe it's necessary. This is much more readable to me, but feel free to disagree. Also check that the code is actually correct, haven't tested it

levnikmyskin commented on src/VirtualDeskManager.cpp:
```diff
@@ -252,24 +252,66 @@ void VirtualDeskManager::resetVdesk(const std::string& arg) {
+    return currentId;
 }
 
 int VirtualDeskManager::nextDeskId(bool cycle) {
```
same comment as for the other function, here you need to keep the vdesksmap.contains call though. In general, though, there's too much repeated code between the two functions, maybe we could simplify this?

levnikmyskin reviewed (changes requested):
Hi, thank you very much for your contribution. I highlighted a few changes that should be made imo.

Ps. please revert the changes to the CI, thank you. We need to build from source for the dev branch

levnikmyskin commented:
also, please update the README with this option

givani30 committed:
Revert "refactor(ci): Simplify CI by installing Hyprland from packages"

This reverts commit 3b77d9e606125f126f27ebc2e26cce3e50570890.

givani30 committed:
refactor: simplify cycle_populated_only implementation

Replace complex do-while loops with clean STL-based filtering approach.
Current desk always included in navigation for intuitive behavior.
Addresses code review feedback with significantly simplified logic.

levnikmyskin committed:
refactor: improve new populated only logic

levnikmyskin commented:
Sorry, didn't like the new changes either. Can you test my latest changes? Let me know if it works the way  you intended this to

givani30 commented:
Thanks for the refactor! I've tested the functionality and it works well in normal scenarios. I
  really appreciate the cleaner structure with the `SCycling` struct and `getCyclingInfo()` helper,
  it makes the algorithm state much more organized. It is also more efficient than the list approach I implemented.

  However, after testing, I've identified some concerns with the current implementation:

  **Potential Issues:**
  1. **Crash risk**: `std::ranges::min(keys)` in getCyclingInfo will crash if `vdesksMap` is ever empty (no defensive
  check)
  2. **Edge case behavior**: When `cycle_populated_only=1` and the current desk becomes unpopulated,
  it's not included in the search candidates, which can lead to unexpected cycling behavior
  3. **Asymmetric logic**: Forward/backward cycling have different fallback paths, making behavior
  less predictable.
 4. **Bug in nextDeskId**: In `nextDeskId` when `populatedOnly=false` and cycling is disabled, if you're
  on the last desktop, it returns `maxId + 1` (an invalid desktop ID) instead of staying on the
  current desktop.

Your refactored version is better  in structural organization and performance. The `SCycling` struct
  makes the algorithm state clear, early termination provides better efficiency, and memory usage is
  lower. My original approach prioritized robustness and safety, it includes defensive
  programming for edge cases, always ensures the current desk is available as a fallback, and uses
  symmetric logic for both directions to reduce the possibility of errors. Both approaches have
  merit, which is why I think combining them would be ideal.

Would you be open to a "best of both worlds" approach? I can create a version that combines your structural improvements with safer edge case handling:

  - Keep the `SCycling` struct organization you introduced
  - Add defensive programming for empty maps and edge cases
  - Ensure current desk is always a valid fallback option
  - Consolidate cycling logic between both nextDeskId and prevDeskId

  This would give us the clean structure you've created while maintaining robustness for edge cases.
  What do you think?

levnikmyskin commented:
Hey, I'm on a holiday and will come back to this PR in 2/3 weeks

givani30 committed:
feat: add cycle_populated_only option with unified cycling logic

- Add cycle_populated_only configuration option to skip empty desktops
- Replace separate nextDeskId/prevDeskId implementations with unified cycleDeskId function
- Fix parameter naming inconsistency in nextDeskId (backwardCycle -> cycle)
- Prevent creation of duplicate empty desks when cycling forward
- Handle sparse virtual desktop IDs correctly (e.g., 1, 3, 5, 7)
- Maintain all existing functionality while simplifying codebase

givani30 committed:
fix: remove static qualifier from PCYCLE_POPULATED_ONLY in isPopulatedOnlyEnabled due to potential race conditions

levnikmyskin commented on .github/actions/setup_base/action.yml:
why these changes?

levnikmyskin commented on src/VirtualDeskManager.cpp:
```diff
@@ -265,53 +266,74 @@ bool VirtualDeskManager::isDeskPopulated(int vdeskId) {
-    // Read config: cycle_populated_only
-    static auto* const PCYCLE_POPULATED_ONLY = (Hyprlang::INT* const*)HyprlandAPI::getConfigValue(PHANDLE, CYCLE_POPULATED_ONLY_CONF)->getDataStaticPtr();
-    bool populatedOnly = **PCYCLE_POPULATED_ONLY;
+bool VirtualDeskManager::isPopulatedOnlyEnabled() {
+    auto* const PCYCLE_POPULATED_ONLY = (Hyprlang::INT* const*)HyprlandAPI::getConfigValue(PHANDLE, CYCLE_POPULATED_ONLY_CONF)->getDataStaticPtr();
+    return **PCYCLE_POPULATED_ONLY;
+}
```
I don't see why this should be necessary honestly. Especially the static modifier shouldn't be removed, see here https://wiki.hypr.land/Plugins/Development/Advanced/#using-the-config

givani30 replied:
You are correct about the performance considerations of the static modifier.
I removed it because of a race condition which I observed where the initial config value was not being read correctly on startup of Hyprland. Because of this, the cycle populated flag was being ignored and the plugin was reverting to the default behavior (false).

Would moving the call inside of the cycleDeskId method be acceptable? This way the race condition is avoided, and the static modifier can be added again.

levnikmyskin commented on src/VirtualDeskManager.cpp:
```diff
@@ -265,53 +266,74 @@ bool VirtualDeskManager::isDeskPopulated(int vdeskId) {
-                candidateId = minId;
-            else
-                return maxId;
+int VirtualDeskManager::cycleDeskId(bool forward, bool allowCycle) {
```
where do we get the `allowCycle` from? The code here got a little too long wrt before

givani30 replied:
The `allowCycle` parameter originates from the different dispatcher used for cycling between virtual desktops
- For `prevdesk` and `nextdesk`, it is set to `false`, which stops cycling at the beginning or end
- `cycleprev` and `cyclenext` pass 'true', which allows wraparound

levnikmyskin replied:
ah sorry, got confused, didn't see that the boolean is passed from `prevDeskId` and `nextDeskId`

levnikmyskin commented:
hi, so let me reply to your points. Let me also first ask you though, if possible, to avoid abusing LLMs, both when writing the code and the replies, as I think this is creating too much noise honestly. Coming to your points:

>  - Crash risk: std::ranges::min(keys) in getCyclingInfo will crash if vdesksMap is ever empty (no defensive
    check)  
but can `vdesksMap` really be empty? It's fine to add the check imo, doesn't hurt, but I'm not sure this edge case can even happen;

>    - Edge case behavior: When cycle_populated_only=1 and the current desk becomes unpopulated,
    it's not included in the search candidates, which can lead to unexpected cycling behavior
This one I haven't understood. The current desk should not matter at all when you're cycling right? What kind of unexpected behaviour does this lead to?

>    - Asymmetric logic: Forward/backward cycling have different fallback paths, making behavior
    less predictable.
That is true, but this is how it was working before too. When we go to the next desk, we go to the next desk even if it doesn't exist (creating it on the fly), unless we're cycling (and we go back to the first desk). When we go to the previous desk, and we're cylcing, we go back to the max available desk. If we're not cycling we don't do anything.

>    - Bug in nextDeskId: In nextDeskId when populatedOnly=false and cycling is disabled, if you're
    on the last desktop, it returns maxId + 1 (an invalid desktop ID) instead of staying on the
    current desktop.
this is the correct behaviour: if cycling is disabled, it means that we're not cycling but we just go to the next desk ID, which is exactly what the function is supposed to do (hence its name)

givani30 commented:
Hi, 
Thank you for the response and the review. I have personally been running the current version so far on my Hyprland setup and it has been running without errors so far. To discuss your points:


> 
> > * Crash risk: std::ranges::min(keys) in getCyclingInfo will crash if vdesksMap is ever empty (no defensive
> >   check)
> 
> but can `vdesksMap` really be empty? It's fine to add the check imo, doesn't hurt, but I'm not sure this edge case can even happen;
> 
While indeed unlikely to happen, this edge case could possibly happen during startup or shutdown of the plugin. Also, if in the future a function is implemented like `reset_all_desktops` or similar which removes all virtual desktops, I can imagine the `std::ranges::min` call being called on an empty map, causing a Hyprland crash.
> > * Edge case behavior: When cycle_populated_only=1 and the current desk becomes unpopulated,
> >   it's not included in the search candidates, which can lead to unexpected cycling behavior
> 
> This one I haven't understood. The current desk should not matter at all when you're cycling right? What kind of unexpected behaviour does this lead to?

Here's a specific scenario where the issue can occur.
1. Imagine `cycle_populated_only=true` and only desk 3 has windows.
2. The user is on desk 3 and closes the last window, making the list of populated desks empty.
3. If we now try to cycle with or `prevdesk` for example, the expected behaviour is to stay on desk 3.
4. The search loop would run through all the desks and find no valid, populated candidates to switch to, making the function finish without a return statement.

My current implementation includes a fallback to `return cycle.currentId`, ensuring predictable behaviour.
> 
> > * Asymmetric logic: Forward/backward cycling have different fallback paths, making behavior
> >   less predictable.
> 
> That is true, but this is how it was working before too. When we go to the next desk, we go to the next desk even if it doesn't exist (creating it on the fly), unless we're cycling (and we go back to the first desk). When we go to the previous desk, and we're cycling, we go back to the max available desk. If we're not cycling we don't do anything.
> 
> > * Bug in nextDeskId: In nextDeskId when populatedOnly=false and cycling is disabled, if you're
> >   on the last desktop, it returns maxId + 1 (an invalid desktop ID) instead of staying on the
> >   current desktop.
> 
> this is the correct behaviour: if cycling is disabled, it means that we're not cycling but we just go to the next desk ID, which is exactly what the function is supposed to do (hence its name)

You are correct, I misunderstood the original behavior and the design intent for asymmetric cycling. I will modify the implementation to match the original logic,  keeping `nextDeskId` and `prevDeskId` separate.

I do have one final question regarding the `nextdesk` behaviour:

What is the intended behavior when you are on the last desk (N), which is empty, and the `nextdesk` dispatcher is called (i.e., `populatedOnly=false` and `cycle=false`)?

Should the plugin continue to create new desks (N+1, N+2, etc.) each time `nextdesk` is called, even if the current desk is unpopulated? Or, if the current desk is already the last one and unpopulated, should it simply stay on the current desk without creating new ones? I feel the latter might be a more intuitive user experience to 
prevent infinitely creating empty desks.

givani30 committed:
refactor: remove isPopulatedOnlyEnabled method and directly access PCYCLE_POPULATED_ONLY in cycleDeskId

levnikmyskin commented:
> Here's a specific scenario where the issue can occur.
> 
>     Imagine cycle_populated_only=true and only desk 3 has windows.
>     The user is on desk 3 and closes the last window, making the list of populated desks empty.
>     If we now try to cycle with or prevdesk for example, the expected behaviour is to stay on desk 3.
>     The search loop would run through all the desks and find no valid, populated candidates to switch to, making the function finish without a return statement.
> 
> My current implementation includes a fallback to return cycle.currentId, ensuring predictable behaviour.

This is not really true, a function annotated to return an integer, will return an integer. Specifically, in my previous implementation and in the scenario you described, what would happen is that:
```c++
    while (cycle.candidateId >= cycle.minId) {
        if (!populatedOnly || isDeskPopulated(cycle.candidateId))
            return cycle.candidateId;
        cycle.candidateId--;
    }
```
the code above would not find any candidate, as you correctly pointed out. Then, this happens:
```c++
    if (populatedOnly) {
        if (backwardCycle && isDeskPopulated(cycle.maxId))
            return cycle.maxId;
        return cycle.currentId;
    }
```
So, since there's no other desk, and nothing is populated, `cycle.currentId` is returned: i.e., the current desk. Behavior seems correct to me.


> What is the intended behavior when you are on the last desk (N), which is empty, and the nextdesk dispatcher is called (i.e., populatedOnly=false and cycle=false)?
>
>Should the plugin continue to create new desks (N+1, N+2, etc.) each time nextdesk is called, even if the current desk is unpopulated?

Yes, if you call `nextdesk` you will just keep going to the next desk. This is how it works now, so this shouldn't be changed unless there are good reasons to do so.

>  I feel the latter might be a more intuitive user experience to prevent infinitely creating empty desks.

This might also be true, but I have two main objections:
1. as I said, this is the current behaviour;
2. if you are on an empty desk and you call the dispatcher to go to the next desk...well, then you go to the next desk. This doesn't seem confusing to me...why else would you call the dispatcher otherwise? :thinking: 


Ps. I just want to stress out that I'm happy to receive contributions, and all of these comments do not mean bad, but as the owner and one of the two maintainers of the plugin I need to check what comes in and what goes out, hope you can understand.

levnikmyskin commented:
implemented in #135
